README.md
<p align="center">
    <a href="https://sitehost.nz" target="_blank">
        <img src=".github/sitehost-logo.svg" height="130">
    </a>
</p>

# SiteHost - Ansible Collection

Requirements
------------

- [Python](https://www.python.org/downloads/) 3.9+
- [Requests](https://pypi.org/project/requests/) 2.28.1
- [Ansible](https://pypi.org/project/ansible/) 7.1.0
- [Ansible Core](https://pypi.org/project/ansible-core/) 2.14.1 _included in ansible package_

## Setup

### Install dependencies

First you will need to install a compatible version of python as defined above. Then you can install the depency with `pip`. 

```bash
pip install ansible
pip install requests
```

### Install the Ansible Collection
You can use `ansible-galaxy` to install the SiteHost Ansible Collection:
```bash
ansible-galaxy collection install sitehost.cloud
```

## Usage
Once the Collection is installed, you can reference it by its [Fully Qualified Collection Namespace (FQCN)](https://github.com/ansible-collections/overview#terminology): `sitehost.cloud.module_name`. 

In order to use this collection, the `SH_API_KEY` and `SH_CLIENT_ID` environment variable must be set to your SiteHost api key and client id respectively. Alternatively, you can pass your api key and client id to the `api_key` and `api_client_id` option of the modules you are calling.  

Please ensure that the appropriate module permission is granted to the API key to use the relevant Ansible module.

#### Example Playbook
```yml
- name: create a 1 core VPS with ubuntu jammy image
  sitehost.cloud.server:
    label: myserver
    location: AKLCITY
    product_code: XENLIT
    image: ubuntu-jammy-pvh.amd64
    api_client_id: "{{ CLIENT_ID }}"
    api_key: "{{ USER_API_KEY }}"
    state: present
```

## Modules
|Name| description|
|----|------------|
|[sitehost.cloud.server](./docs/server.md)| Manages SiteHost virtual server|
|[sitehost.cloud.dns](./docs/dns.md)| Manages SiteHost DNS records and zones|
|[sitehost.cloud.stack](./docs/dns.md)| Manages SiteHost Cloud Containers|


CHANGELOG.rst
============================
Sitehost.Cloud Release Notes
============================

.. contents:: Topics


v1.2.0
======

Minor Changes
-------------

- sitehost.cloud.server - Now supports provisioning Cloud Container Servers.

New Modules
-----------

- sitehost.cloud.stack - Manages Cloud Containers

v1.1.0
======

Bugfixes
--------

- Fixed an issue with server module when state is `started` or `stopped`.

New Modules
-----------

- sitehost.cloud.dns - Manage DNS zones and records

v1.0.0
======

Release Summary
---------------

Initial release of the collection.


plugins/modules/dns.py
#!/usr/bin/python

# Copyright: (c) 2018, Terry Jones <terry.jones@example.org>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
from __future__ import absolute_import, division, print_function

__metaclass__ = type

DNS_TYPE_LIST = ["A", "AAAA", "CNAME", "MX", "TXT", "CAA", "SRV"]
HTTP_GET = "GET"
HTTP_POST = "POST"

DOCUMENTATION = r"""
---
module: dns

short_description: Manage DNS zones and records

version_added: "1.1.0"

description: Used for creating, deleting and updating DNS records and DNS zones on your SiteHost accout.

author:
  - "SiteHost Developers (developers@sitehost.co.nz)"

options:
    state:
        description: 
            - Desired state of the DNS zone / record.
            - C(present) will either upgrade or create a DNS zone or record.
            - C(absent) will delete a DNS zone or record.
        required: true
        type: str
    domain:
        description:
            - The selected DNS zone.
            - If DNS zone does not exist when adding a new DNS record, it will be automatically created.
        required: True
        type: str
        aliases: 
            - zone
            - dns_zone
    record_id:
        description:
            - The id of an exisiting DNS record to update or delete.
        type: int
    name:
        description:
            - The host name, alias, or service being defined by the record.
            - Please use the FQDN instead of relative names like C(@) or C(www)
        type: str
    type:
        description:
            - The type of record you would like to use.
        choices: [A,AAAA,CNAME,MX,TXT,CAA,SRV]
        type: str
    priority:
        description:
            - The priority of the host for C(SRV,MX) records.
        type: int
    content:
        description:
            - This is the value of the record, depending on the record type.
        type: str
"""

EXAMPLES = r"""
# Create a DNS record and print the record id
- name: create an A dns record for @
    sitehost.cloud.dns:
        domain: mydomain.co.nz
        type: A
        name: mydomain.co.nz
        content: 255.255.255.255
        api_client_id: "{{ CLIENT_ID }}"
        api_key: "{{ USER_API_KEY }}"
        state: present
    register: recordouput

- name: print the record id
    ansible.builtin.debug:
        var: recordouput.dns.id

# Update a previously created dns record
- name: update previously created record
    sitehost.cloud.dns:
        domain: mydomain.co.nz
        record_id: 1234567
        type: A
        name: new.mydomain.co.nz
        content: 255.255.255.254
        api_client_id: "{{ CLIENT_ID }}"
        api_key: "{{ USER_API_KEY }}"
        state: present

# Delete a previously created DNS record
- name: Delete dns record
    sitehost.cloud.dns:
        domain: mydomain.co.nz
        record_id: 1234567
        api_client_id: "{{ CLIENT_ID }}"
        api_key: "{{ USER_API_KEY }}"
        state: absent

# create a DNS zone without adding any records
- name: create DNS zone
    sitehost.cloud.dns:
        domain: newdomain.com
        api_client_id: "{{ CLIENT_ID }}"
        api_key: "{{ USER_API_KEY }}"
        state: present

# delete a DNS zone
- name: delete DNS zone
    sitehost.cloud.dns:
        domain: newdomain.com
        api_client_id: "{{ CLIENT_ID }}"
        api_key: "{{ USER_API_KEY }}"
        state: absent
"""

RETURN = r"""
dns:
    description: The dns record being created or modified.
    type: dict
    returned: success
    sample: {
        "change_date": "1695252962",
        "content": 255.255.255.255",
        "id": "1234567",
        "name": "www.mydomain.co.nz",
        "prio": "0",
        "state": "0",
        "ttl": "3600",
        "type": "A"
    }
msg:
    description: A short messages showing the state of the module execution.
    type: str
    returned: always
    sample: "DNS record created with id: 1234567"
"""

from collections import OrderedDict  # noqa: E402

from ansible.module_utils.basic import AnsibleModule  # noqa: E402

from ..module_utils.sitehost import SitehostAPI  # noqa: E402


class AnsibleSitehostDNS:
    def __init__(self, module, api):
        self.sh_api = api
        self.module = module
        self.result = {
            "changed": False,
            "dns": dict(),
        }

    def update_or_add(self):
        """Check if the current dns record needs to be updated or a new dns zone
        or record needs to be created when state is set to present."""
        if self.module.params["record_id"]:
            self.update_dns_record()
        elif not self.module.params["name"]:
            self.create_domain()
        else:
            self.add_dns_record()

    def absent(self):
        """Check if the DNS record or zone needs to be deleted when the state
        is set to absent."""
        if self.module.params["record_id"]:
            self.delete_dns_record()
        else:
            self.delete_domain()

    def update_dns_record(self):
        """Update an existing DNS record."""
        #  check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        if not self._get_domain():
            self.module.fail_json(msg="ERROR: DNS zone does not exist.")
        if not self._get_record_by_id():
            self.module.fail_json(msg="ERROR: DNS Record does not exist.")

        # update the DNS record
        body = OrderedDict()
        body["domain"] = self.module.params["domain"]
        body["record_id"] = self.module.params["record_id"]
        body["type"] = self.module.params["type"]
        body["name"] = self.module.params["name"]
        body["content"] = self.module.params["content"]
        body["prio"] = self.module.params["priority"]

        update_result = self.sh_api.api_query(
            path="/dns/update_record.json", method=HTTP_POST, data=body
        )

        if "status" in update_result and not update_result["status"]:  # update failed due to incorrect parameters
            self.module.fail_json(msg=update_result["msg"])

        self.result["msg"] = "DNS record successfully updated"
        self.result["dns"] = self._get_record_by_id()
        self.result["changed"] = True
        self.module.exit_json(**self.result)

    def add_dns_record(self):
        """Adds a DNS record, will create DNS zone if it does not exist."""
        # check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        # create DNS zone if it does not exist
        if self._get_domain() is None:
            self._create_zone(self.module.params["domain"])

        # create the DNS record
        body = OrderedDict()
        body["domain"] = self.module.params["domain"]
        body["type"] = self.module.params["type"]
        body["name"] = self.module.params["name"]
        body["content"] = self.module.params["content"]

        add_result = self.sh_api.api_query(
            path="/dns/add_record.json", method=HTTP_POST, data=body
        )
        if "status" in add_result and not add_result["status"]:  # adding failed due to incorrect parameters
            self.module.fail_json(msg=add_result["msg"])

        # get the newly created DNS record to show in output of module
        record_list = self.sh_api.api_query(
            path="/dns/list_records.json",
            method=HTTP_GET,
            query_params={"domain": self.module.params["domain"]},
        )
        if "return" not in record_list:
            self.module.fail_json(msg=record_list)

        return_record_list = record_list["return"]
        return_record_list = filter(
            lambda x: x["name"] == self.module.params["name"], return_record_list
        )
        new_record = max(return_record_list, key=lambda x: int(x["change_date"]))

        self.result["msg"] = f"DNS record created with id: {new_record['id']}"
        self.result["dns"] = new_record
        self.result["changed"] = True
        self.module.exit_json(**self.result)

    def delete_domain(self):
        """Deletes a DNS zone."""
        # check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        # check if domain zone exists
        if self._get_domain() is None:
            self.module.exit_json(msg="Specified DNS zone does not exist")

        # delete the zone
        body = OrderedDict()
        body["domain"] = self.module.params["domain"]

        delete_result = self.sh_api.api_query(path="/dns/delete_domain.json", method=HTTP_POST, data=body)
        if "status" in delete_result and not delete_result["status"]:
            self.module.fail_json(msg=delete_result)

        self.module.exit_json(msg="DNS zone deleted", changed=True)

    def create_domain(self):
        """Creates a new DNS zone only when there is no DNS records specified."""
        if self._get_domain():
            self.module.exit_json(msg="DNS zone already exist")

        # check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        create_result = self._create_zone()

        if "status" in create_result and not create_result["status"]:
            self.module.fail_json(msg=create_result["msg"])

        self.result["msg"] = f"DNS zone \"{self.module.params['domain']}\" created"
        self.result["changed"] = True

        self.module.exit_json(**self.result)

    def delete_dns_record(self):
        """Deletes a DNS record."""
        # check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        if self._get_domain() is None:
            self.module.fail_json(msg="ERROR: DNS zone does not exist.")
        if self._get_record_by_id() is None:
            self.module.exit_json(msg="DNS Record does not exist.", changed=False)

        # delete the dns record
        body = OrderedDict()
        body["domain"] = self.module.params["domain"]
        body["record_id"] = self.module.params["record_id"]
        self.sh_api.api_query(
            path="/dns/delete_record.json", method=HTTP_POST, data=body
        )

        self.result["msg"] = "DNS record deleted"
        self.result["changed"] = True
        self.module.exit_json(**self.result)

    def _get_record_by_id(self, record_id=None):
        """
        Get the DNS record by ID.

        :param record_id: The dns record to retrieve. If it is not set,
                    then use `record_id` parameter in playbook.
        :returns: The DNS record if it exists, otherwise return none.
        :rtype: list or None
        """
        if record_id is None:
            record_id = self.module.params["record_id"]

        retrieved_records = self.sh_api.api_query(
            path="/dns/list_records.json",
            method=HTTP_GET,
            query_params=OrderedDict({"domain": self.module.params["domain"]}),
        )
        dns_record = [
            record
            for record in retrieved_records["return"]
            if record["id"] == str(record_id)
        ]

        return dns_record if dns_record else None

    def _get_domain(self, zone=None):
        """
        Get the DNS zone given by domain parameters.

        :param zone: The dns zone to retrieve. If it is not set,
                    then use `domain` parameter in playbook.
        :returns: The Zone information if it exists, otherwise
                    return none
        :rtype: dict or None
        """
        if zone is None:
            zone = self.module.params["domain"]

        body = OrderedDict()
        body["query[domain]"] = zone
        retrieved_zone = self.sh_api.api_query(
            path="/dns/search_domains.json", method=HTTP_POST, data=body
        )["return"]

        return retrieved_zone if retrieved_zone else None

    def format_parameters(self):
        """ensures that the parameters are lower case"""
        if self.module.params["domain"]:
            self.module.params["domain"] = self.module.params["domain"].lower()
        if self.module.params["name"]:
            self.module.params["name"] = self.module.params["name"].lower()

    def _create_zone(self, domain=None):
        """
        Create a DNS zone.

        :params domain: Specifies the DNS zone to create. If it is not set,
                    then use `domain` parameter in playbook
        :returns: The api output.
        :rtype: dict
        """
        # check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        if domain is None:
            domain = self.module.params["domain"]

        body = OrderedDict()
        body["domain"] = domain
        api_result = self.sh_api.api_query(
            path="/dns/create_domain.json", method=HTTP_POST, data=body
        )

        return api_result


def main():
    argument_spec = SitehostAPI.sitehost_argument_spec()
    argument_spec.update(
        dict(
            domain=dict(type="str", required=True, aliases=["zone", "dns_zone"]),
            record_id=dict(type="int"),
            name=dict(type="str"),
            type=dict(
                type="str",
                choices=DNS_TYPE_LIST,
            ),
            priority=dict(type="int"),
            content=dict(type="str"),
            state=dict(
                type="str",
                choices=["present", "absent"],
                default="present",
            ),
        )
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        required_together=("name", "type", "content"),
        supports_check_mode=True,
    )

    sitehost_api = SitehostAPI(
        module=module,
        api_key=module.params["api_key"],
        api_client_id=module.params["api_client_id"],
    )

    sitehost_dns = AnsibleSitehostDNS(module=module, api=sitehost_api)

    sitehost_dns.format_parameters()

    state = module.params["state"]

    if state == "present":
        sitehost_dns.update_or_add()
    else:  # state is absent
        sitehost_dns.absent()


if __name__ == "__main__":
    main()


plugins/modules/stack.py
#!/usr/bin/python

# Copyright: (c) 2018, Terry Jones <terry.jones@example.org>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
from __future__ import absolute_import, division, print_function

__metaclass__ = type

HTTP_GET = "GET"
HTTP_POST = "POST"
SCHEDULER_JOB_TYPE = "scheduler"

DOCUMENTATION = r"""
---
module: stack

version_added: "1.2.0"

short_description: Manages Cloud Containers

description: Used for creating, deleting, updating, starting, and stopping Cloud Containers on your SiteHost account.

author:
  - "SiteHost Developers (developers@sitehost.co.nz)"

options:
    state:
        description:
            - Indicates the desired state of the Cloud Container.
            - C(present) will either update or create a Cloud Container.
            - C(absent) will delete a Cloud Container.
            - C(started) for powering on the container.
            - C(stopped) for powering off the container.
            - C(restarted) for restarting the container.
        default: present
        choices: [present, absent, started, stopped, restarted]
        type: str
    server:
        description: 
            - The Cloud Container server to operate on.
        required: true
        type: str
    name:
        description:
            - A unique Hash assigned to the server
            - L(Generate, https://docs.sitehost.nz/api/v1.2/?path=/cloud/stack/generate_name&action=GET) one with the API before hand before using it.
        required: false
        type: str
    label:
        description: 
            - User chosen label of the Container.
            - The label format must be a valid FQDN.
        required: false
        type: str
    docker_compose:
        description:
            - The docker_compose file that needs to be set when creating a server.
            - Check out the documentation in the L(SiteHost Ansible Github repo,https://github.com/sitehostnz/sitehost_ansible/blob/main/docs/stack.md) to learn more about setting up a docker_compose file for Cloud Containers.
        required: false
        type: yaml
    api_key:
        description:
            - Your SiteHost api key L(generated from CP,https://kb.sitehost.nz/developers/api#creating-an-api-key).
        required: true
        type: str
    api_client_id:
        description:
            - The client id of your SiteHost account.
        required: true
        type: int

"""

EXAMPLES = r"""
# create a Cloud Container running apache + php 8.2
- name: create a container
  sitehost.cloud.stack:
    server: ch-mycloudse
    name: ccb7a31da52e5b47
    label: mycontainer.co.nz
    docker_compose: |
        version: '2.1'
        services:
            ccb7a31da52e5b47:
                container_name: ccb7a31da52e5b47
                environment:
                    - 'VIRTUAL_HOST=mycontainer.co.nz,www.mycontainer.co.nz'
                    - CERT_NAME=mycontainer.co.nz
                expose:
                    - 80/tcp
                image: 'registry.sitehost.co.nz/sitehost-php82-apache:4.0.1-jammy'
                labels:
                    - 'nz.sitehost.container.website.vhosts=mycontainer.co.nz,www.mycontainer.co.nz'
                    - nz.sitehost.container.image_update=True
                    - nz.sitehost.container.label=mycontainer.co.nz
                    - nz.sitehost.container.type=www
                    - nz.sitehost.container.monitored=True
                    - nz.sitehost.container.backup_disable=False
                restart: unless-stopped
                volumes:
                    - '/data/docker0/www/ccb7a31da52e5b47/crontabs:/cron:ro'
                    - '/data/docker0/www/ccb7a31da52e5b47/application:/container/application:rw'
                    - '/data/docker0/www/ccb7a31da52e5b47/config:/container/config:ro'
                    - '/data/docker0/www/ccb7a31da52e5b47/logs:/container/logs:rw'
        networks:
            default:
                external:
                    name: infra_default

    api_client_id: "{{ CLIENT_ID }}"
    api_key: "{{ USER_API_KEY }}"
    state: present

# delete a Cloud Container
- name: delete a container
  sitehost.cloud.stack:
    server: ch-mycloudse
    name: ccb7a31da52e5b47
    api_client_id: "{{ CLIENT_ID }}"
    api_key: "{{ USER_API_KEY }}"
    state: absent

# powering up a Cloud Container
- name: start container
  sitehost.cloud.stack:
    server: ch-mycloudse
    name: ccb7a31da52e5b47
    api_client_id: "{{ CLIENT_ID }}"
    api_key: "{{ USER_API_KEY }}"
    state: started

"""

RETURN = r"""
# These are examples of possible return values, and in general should use other names for return values.
msg:
    description: A short messages showing the state of the module execution.
    type: str
    returned: always
    sample: 'Container ccb7a31da52e5b47 created'
stack:
    description: The Cloud Container status.
    type: dict
    returned: success
    sample: {
        "client_id": "1234567",
        "containers": [
            {
                "backups": true,
                "container_id": "b3a1775335a7e9b9c85c835fa1a5973b19e67e7a7a4577121f8e879ffafecf80",
                "date_added": "2023-10-11 15:14:52",
                "date_updated": "2023-10-11 15:14:59",
                "image": "registry.sitehost.co.nz/sitehost-php82-apache:4.0.1-jammy",
                "is_missing": "0",
                "monitored": true,
                "name": "ccb7a31da52e5b47",
                "pending": null,
                "size": "0",
                "ssl_enabled": false,
                "state": "Up"
            }
        ],
        "date_added": "2023-10-11 15:14:52",
        "date_updated": "2023-10-11 15:14:59",
        "docker_file": "version: '2.1'\nservices:\n    ccb7a31da52e5b47:\n        container_name: ccb7a31da52e5b47\n        environment:\n            - 'VIRTUAL_HOST=mycontainer.co.nz,www.mycontainer.co.nz'\n            - CERT_NAME=mycontainer.co.nz\n        expose:\n            - 80/tcp\n        image: 'registry.sitehost.co.nz/sitehost-php82-apache:4.0.1-jammy'\n        labels:\n            - 'nz.sitehost.container.website.vhosts=mycontainer.co.nz,www.mycontainer.co.nz'\n            - nz.sitehost.container.image_update=True\n            - nz.sitehost.container.label=mycontainer.co.nz\n            - nz.sitehost.container.type=www\n            - nz.sitehost.container.monitored=True\n            - nz.sitehost.container.backup_disable=False\n        restart: unless-stopped\n        volumes:\n            - '/data/docker0/www/ccb7a31da52e5b47/crontabs:/cron:ro'\n            - '/data/docker0/www/ccb7a31da52e5b47/application:/container/application:rw'\n            - '/data/docker0/www/ccb7a31da52e5b47/config:/container/config:ro'\n            - '/data/docker0/www/ccb7a31da52e5b47/logs:/container/logs:rw'\nnetworks:\n    default:\n        external:\n            name: infra_default\n",
        "ip_addr_server": "255.255.255.255",
        "is_missing": "0",
        "label": "mycontainer.co.nz",
        "name": "ccb7a31da52e5b47",
        "pending": null,
        "server_id": "12345",
        "server_label": "my cloud server",
        "server_name": "ch-mycloudse",
        "server_owner": true
    }

"""

from collections import OrderedDict  # noqa: E402

from ansible.module_utils.basic import AnsibleModule  # noqa: E402

from ..module_utils.sitehost import SitehostAPI  # noqa: E402


class AnsibleSitehostStack:
    _STACK_STATE_MAP = {"Up": "started", "Exit 0": "stopped"}
    # converts module user input to appropriate api call state
    _API_STATE_MAP = {"started": "start", "stopped": "stop", "restarted": "restart"}

    def __init__(self, module, api):
        self.sh_api = api
        self.module = module
        self.result = {
            "changed": False,
            "stack": dict(),
        }

    def create_or_update(self):
        if self._get_stack():
            # If the container already exists, update its configuration.
            self.update_stack()
        else:
            # otherwise create a new one.
            self.create_stack()

    def create_stack(self):
        """create a Cloud Container."""
        #  check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        body = OrderedDict()
        body["server"] = self.module.params["server"]
        body["name"] = self.module.params["name"]
        body["label"] = self.module.params["label"]
        body["enable_ssl"] = 0
        body["docker_compose"] = self.module.params["docker_compose"]

        # creates Cloud Container
        api_result = self.sh_api.api_query(
            path="/cloud/stack/add.json",
            method=HTTP_POST,
            data=body,
        )

        self.sh_api.wait_for_job(
            job_id=api_result["return"]["job_id"], job_type=SCHEDULER_JOB_TYPE
        )

        self.result["msg"] = f"Container {self.module.params['name']} created"
        self.result["stack"] = self._get_stack()
        self.result["changed"] = True

        self.module.exit_json(**self.result)

    def update_stack(self):
        """Updates a Cloud Container. This method updates the label and/or
        docker_compose of an existing container based on the provided parameters.
        """
        #  check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        body = OrderedDict()
        body["server"] = self.module.params["server"]
        body["name"] = self.module.params["name"]

        if self.module.params.get("label"):  #  update label if defined
            body["params[label]"] = self.module.params["label"]
            self.result["changed"] = True
        if self.module.params.get(
            "docker_compose"
        ):  #  update docker_compose if defined
            body["params[docker_compose]"] = self.module.params["docker_compose"]
            self.result["changed"] = True

        api_result = self.sh_api.api_query(
            path="/cloud/stack/update.json", method=HTTP_POST, data=body
        )

        # if a scheduler job is created, wait for it
        if api_result.get("return"):
            self.sh_api.wait_for_job(
                job_id=api_result["return"]["job_id"], job_type=SCHEDULER_JOB_TYPE
            )

        self.result["msg"] = f"Container {self.module.params['name']} updated"
        self.result["stack"] = self._get_stack()

        self.module.exit_json(**self.result)

    def delete_stack(self):
        """Deletes a Cloud Container."""
        #  check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        body = OrderedDict()
        body["server"] = self.module.params["server"]
        body["name"] = self.module.params["name"]

        # if container does not exist, then skip the task
        if self._get_stack() is None:
            self.module.exit_json(
                msg=f"Container {self.module.params['name']} does not exist",
                changed=False,
            )

        # deletes the Cloud Container
        api_result = self.sh_api.api_query(
            path="/cloud/stack/delete.json",
            method=HTTP_POST,
            data=body,
        )

        self.sh_api.wait_for_job(
            job_id=api_result["return"]["job_id"], job_type=SCHEDULER_JOB_TYPE
        )

        self.result["msg"] = f"Container {self.module.params['name']} deleted"
        self.result["changed"] = True

        self.module.exit_json(**self.result)

    def _get_stack(self, container_to_check=None):
        """
        Get Cloud Container information.

        :params container_to_check: select the container to get, if not provided.
        :return: Information about the container. Returns None if
                the container doesn't exist.
        :rtype: dict or None
        """
        if container_to_check is None:
            container_to_check = self.module.params["name"]

        retrived_container = self.sh_api.api_query(
            path="/cloud/stack/get.json",
            method=HTTP_GET,
            query_params=OrderedDict(
                [
                    ("server", self.module.params["server"]),
                    ("name", container_to_check),
                ]
            ),
            skip_status_check=True,
        )

        return retrived_container["return"] if retrived_container["status"] else None

    def handle_power_state(self):
        """Use to start, stop and restart containers"""

        # check if the container exist
        if self._get_stack() is None:
            #  check mode, container might not be created yet
            if self.module.check_mode:
                self.module.exit_json(changed=True)
            self.module.fail_json(msg="ERROR: Specified container does not exist")

        requested_stack_state = self.module.params["state"]

        # otherwise get the current container state to check if task can be skipped
        current_stack_state = self._get_stack()["containers"][0]["state"]

        # the container state is the requested state already, skip task.
        if (
            AnsibleSitehostStack._STACK_STATE_MAP[current_stack_state]
            == requested_stack_state
        ):
            self.result["msg"] = f"Container already {requested_stack_state}"
            self.result["stack"] = self._get_stack()
            self.module.exit_json(**self.result)

        ## requested container state is different from current state

        #  check mode
        if self.module.check_mode:
            self.module.exit_json(changed=True)

        body = OrderedDict()
        body["server"] = self.module.params["server"]
        body["name"] = self.module.params["name"]

        # start or stop container
        api_result = self.sh_api.api_query(
            path=f"/cloud/stack/{AnsibleSitehostStack._API_STATE_MAP[requested_stack_state]}.json",
            method=HTTP_POST,
            data=body,
        )

        self.sh_api.wait_for_job(
            job_id=api_result["return"]["job_id"], job_type=SCHEDULER_JOB_TYPE
        )
        self.result[
            "msg"
        ] = f"Container {self.module.params['name']} {requested_stack_state}"
        self.result["stack"] = self._get_stack()
        self.result["changed"] = True
        self.module.exit_json(**self.result)


def main():
    argument_spec = SitehostAPI.sitehost_argument_spec()
    argument_spec.update(
        server=dict(type="str", required=True),
        name=dict(type="str", required=True),
        label=dict(type="str"),
        docker_compose=dict(type="str"),
        state=dict(
            type="str",
            choices=["present", "absent", "started", "stopped", "restarted"],
            default="present",
        ),
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
    )

    sitehost_api = SitehostAPI(
        module=module,
        api_key=module.params["api_key"],
        api_client_id=module.params["api_client_id"],
    )

    sitehoststack = AnsibleSitehostStack(module=module, api=sitehost_api)

    state = module.params["state"]

    if state == "present":
        sitehoststack.create_or_update()
    elif state == "absent":
        sitehoststack.delete_stack()
    else:
        sitehoststack.handle_power_state()


if __name__ == "__main__":
    main()
